src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,625 @@
+    fprintf(f, "\n");
+    // build .o from .cpp
+    fprintf(f, "$(MODULE).o: $(MODULE).cpp\n");
+    fprintf(f, "\t@echo JAVA_HOME '$(JAVA_HOME)'\n");
+    fprintf(f, "\t@echo JAVA_INCLUDE '$(JAVA_INCLUDE)'\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $@ -c $< 2> $(MODULE).err && \\\n");
+    fprintf(f, "\t  $(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $(MODULE).tmp $(MODULE).o 2>> $(MODULE).err ; \\\n");
+    fprintf(f, "\tstatus=$$? ; \\\n");
+    fprintf(f, "\tif [ $$status -ne 0 ] || [ -z $(MODULE).tmp ]; then \\\n");
+    fprintf(f, "\t  rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  echo FAIL ; exit 1 ; \\\n");
+    fprintf(f, "\tfi\n");
+    fprintf(f, "\t-rm -f $(MODULE).err\n");
+    fprintf(f, "\tperl -e 'rename \"$(MODULE).tmp\", \"$(MODULE).new\"'\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_build() {
+    // Try to create the .new file, we hold the global lock so there is no race
+    int fd = ::open(hm_new_.c_str(), O_WRONLY|O_CREAT|O_TRUNC, 0666);
+    if (fd < 0) {
+      perror("open");
+      assert(false);
+    }
+    ::close(fd);
+    // The .new file may look the same age as the .cpp file, but
+    // the makefile is written to ignore the .new timestamp
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "make -B -C " << config.module_dir_ << " -f " << hm_mak_ << " 1>/dev/null";
+    // run the command synchronously, while holding the big_mutex
+    int rc = system(ss.str().c_str());
+    return (rc == 0);
+  }
+};
+
+NativeModule::NativeModule(
+  const char* options,
+  const char* source,
+  const char* include,
+  bool force_build
+) :
+  build_state_(kInit),
+  load_state_(kInit),
+  key_(hash_two_strings(options, source)),
+  is_global_(false),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  // Master constructor - try to get module built in local file
+  config.ensure_module_dir_exists();
+  bool have_lib = (!force_build && file_exists(lib_name_));
+  if (have_lib) {
+    build_state_ = kPass;
+  } else {
+    // The file doesn't exist, let's build it
+    ModuleBuilder builder(options, source, include, key_);
+    build_state_ = (builder.try_to_build() ? kPass : kFail);
+  }
+}
+
+NativeModule::NativeModule(
+  bool is_global,
+  const char* key,
+  ssize_t binary_size,
+  const void* binary
+) :
+  build_state_(is_global ? kPass : kInit),
+  load_state_(is_global ? kPass : kInit),
+  key_(key),
+  is_global_(is_global),
+  dlopen_handle_(nullptr),
+  lib_name_(config.get_lib_name(key_)),
+  new_name_(config.get_new_name(key_)) {
+  std::lock_guard<std::mutex> mylock(big_mutex);
+  // Worker constructor - try to get the binary written to local file
+  if (is_global_) return;
+  int rc = 0;
+  config.ensure_module_dir_exists();
+  struct stat lib_stat;
+  if (file_stat(lib_name_, &lib_stat) && (lib_stat.st_size == binary_size)) {
```

Comment:
I don't see how the lib can exist at this point.  I feel like you should just write the binary out no questions asked and dispense will the old/unlink/rename/etc.

Reply:
The first case is when a worker thread is running the same process as the master -
in that case the lib file already exists because the master created it.

Also in the (current) shuffle, each thread is calling nativeCtorWorker for each item, and then
close()ing it after 1 item. With no coordination, it's possible for all workers to be without a
worker NativeModule simultaneously, in which case the shared NativeModule would disappear, and then the next nativeCtorWorker would construct a new one, even though the lib file still exists.

But I agree that now the files are per-session, and if we also eliminate force_build, then it will suffice to use the file if it exists (in same process as master, or not the first worker trying to use it), or else write it out (the first worker in a different process from the master).

Reply:
Can't you handle this case by adding the master also add the module to `module_table`?  Then the worker will pickup the module compiled by the master thread.

Reply:
I'm thinking the simple thing to do is to have precisely one NativeModule object corresponding to
each key, keep them in module_table, and never throw them away.  Then in both the master-first and worker-first cases, the constructor for the first NativeModule is responsible for populating the lib-file, and the existence of a NativeModule in the module_table would imply the existence of the lib, eliminating the filesystem checks.

The theoretical downside is that it then becomes much more difficult to ever unload a module
which is no longer being used.  I don't think that's a problem for the current fire-up-a-cluster-for-each-session architecture, it might become an issue in a future one-big-cluster-serves-many-users architecture.

Reply:
I think there's a happy medium.  How about this:

 - one module per key, keep them all in module_table, but with weak_ptrs (so unused modules are unloaded),
 - lock all operations creating lib files,
 - never delete or modify lib files,
 - when you create the module, if the file is already there (checking with the lock), just use it.

This way, we don't synchronize using the file system, reuse compilation throughout the session, and we unload unused modules.

Given that we're letting the module dir grow without bound, I'm not sure how bad it is to leave things loaded. I assume pages of unused modules will just get freed by the buffer cache.

Reply:
Will do.  Note only that there's no requirement for the master module to continue existing
until the first worker is created - and usually it won't, since things are single-threaded and it's most convenient to close() the master as soon as you've got the (key,binary) pair to pass around for the construction of workers.